sklearn/metrics/pairwise.py

```diff
@@ -829,6 +834,172 @@ def cosine_distances(X, Y=None):
+    X_cat = _safe_indexing(X, col_idx, axis=1)
+    X_num = _safe_indexing(X, col_idx, axis=1, complement=True)
+
+    return X_cat, X_num
+
+
+def gower_distances(X, Y=None, categorical_features=None, scale=True,
+                    min_values=None, scale_factor=None):
+    """Compute the distances between the observations in X and Y,
+    that may contain mixed types of data, using an implementation
+    of Gower formula.
+
+    Parameters
+    ----------
+    X : {array-like, pandas.DataFrame} of shape (n_samples, n_features)
+
+    Y : {array-like, pandas.DataFrame} of shape (n_samples, n_features), \
+        default=None
+
+    categorical_features : array-like of str, array-like of int, \
+            array-like of bool, slice or callable, default=None
+        Indexes the data on its second axis. Integers are interpreted as
+        positional columns, while strings can reference DataFrame columns
+        by name.
+        A callable is passed the input data `X` and can return any of the
+        above. To select multiple columns by name or dtype, you can use
+        :obj:`~sklearn.compose.make_column_selector`.
+
+        By default all non-numeric columns are considered categorical.
+
+    scale : bool, default=True
+        Indicates if the numerical columns will be scaled between 0 and 1.
+        If false, it is assumed the numerical columns are already scaled.
+        The scaling factors, _i.e._ min and max, are taken from both ``X`` and
+        ``Y``.
+
+    min_values : ndarray of shape (n_features,), default=None
+        Per feature adjustment for minimum. Equivalent to
+        ``min_values - X.min(axis=0) * scale_factor``
+        If provided, ``scale_factor`` should be provided as well.
+
+    scale_factor : ndarray of shape (n_features,), default=None
+        Per feature relative scaling of the data. Equivalent to
+        ``(max_values - min_values) / (X.max(axis=0) - X.min(axis=0))``
+        If provided, ``min_values`` should be provided as well.
+
+    Returns
+    -------
+    distances : ndarray of shape (n_samples_X, n_samples_Y)
+
+    References
+    ----------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    -----
+    Categorical ordinal attributes should be treated as numeric for the purpose
+    of Gower similarity.
+
+    Current implementation does not support sparse matrices.
+
+    All the non-numerical types (e.g., str), are treated as categorical
+    features.
+
+    This implementation modifies the Gower's original similarity measure in
+    the folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation
+    returns 1-S.
+    """
+    def _nanmanhatan(x, y):
+        return np.nansum(np.abs(x - y))
```

Comment:
Is using this row-wise function call worthwhile relative to something more vectorised?

Reply:
It's kinda easy to do more vectorized when X and Y are of the same size, otherwise I'm not sure if it's worth the complexity.

Reply:
Also, this implementation is significantly faster than the cosine distances for instance. So I don't think we should worry about the speed too much?